Amqp/src/Akka.Streams.Amqp/AmqpConnectionProvider.cs

```diff
@@ -0,0 +1,419 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Akka.Util;
+using RabbitMQ.Client;
+
+namespace Akka.Streams.Amqp
+{
+    /// <summary>
+    /// Only for internal implementations
+    /// </summary>
+    public abstract class AmqpConnectionProvider
+    {
+        public abstract IConnection Get();
+
+        public virtual void Release(IConnection connection)
+        {
+            if(connection.IsOpen)
+                connection.Close();
+        }
+    }
+
+    /// <summary>
+    /// Connects to a local AMQP broker at the default port with no password.
+    /// </summary>
+    public sealed class AmqpLocalConnectionProvider : AmqpConnectionProvider
+    {
+        private AmqpLocalConnectionProvider()
+        {
+            
+        }
+        public static AmqpLocalConnectionProvider Instance=> new AmqpLocalConnectionProvider();
+        public override IConnection Get() => new ConnectionFactory().CreateConnection();
+    }
+
+    public sealed class AmqpUriConnectionProvider : AmqpConnectionProvider
+    {
+        private readonly Uri _uri;
+
+        public static AmqpUriConnectionProvider Create(Uri uri)=> new AmqpUriConnectionProvider(uri);
+
+        public static AmqpConnectionProvider Create(string uri)=> new AmqpUriConnectionProvider(new Uri(uri));
+
+        private AmqpUriConnectionProvider(Uri uri)
+        {
+            _uri = uri;
+        }
+        public override IConnection Get()
+        {
+            var factory = new ConnectionFactory();
```

Comment:
Are you sure it's ok to create a `ConnectionFactory` every time `Get` is called?